atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -202,6 +202,7 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
 
                                 Transaction transaction = createTransaction(
                                         immutableTs, startTimestampSupplier, immutableTsLock, condition);
+                                transaction.onSuccess(() -> lockWatchManager.onSuccess(transaction.getTimestamp()));
```

Comment:
According to the contract of `onSuccess`, this should execute first (unless there are other calls registered while the transaction is being created).

However, if this fails, we'll still clean up the state, but won't flush the values to the central cache, which is acceptable.

Reply:
I agree this is _acceptable_. If I'm not mistaken we will end up, for successful transactions, cleaning up the state twice (`OpenTransactionImpl#finish()` also removes the transaction state from the cache) - do we need to update that?

Reply:
I think calling this twice is perfectly fine. However, if you disagree, we can rewire so it does not.